### DISCUSSION TITLE
Allow references in Folium

### DISCUSSION THREAD

**Discussion by hansthen** [python-visualization/folium#2045] [Ideas]
There are a few issues open that it is impossible to reuse elements such as Icons for layers. Apart from the CustomIcons, other things that are currently impossible are circular references. E.g. I would like to pass a reference to a LayerControl from a Layer and vice versa, so that they can update the other's state during events. This is currently not possible.

The underlying cause of this is the `Branca` object model. `Branca` only allows containment relations. This is fine for modelling HTML or any other document object model. However, it does not much the `Leaflet` object model, which uses references instead of containment. 

The following is a two part proposal on how to make references possible. I will first describe the required changes to the object model. Second I will describe how we can make this proposal work without breaking existing client code. Out of scope will be developers who may have made custom plugins or inherited from our object model outside of the `Folium` project.

**Comment by hansthen** [comment]
**First part, establish new object model**

1. Create a new Branca Element to represent a Javascript code block.
2. Make Folium Leaflet objects no longer recursive.
3. When creating Folium objects add them to the Javascript code block
4. Method invocations on Folium objects that modify the structure (like `FeatureGroup.addTo` and `Map.addTo`) should generate the equivalent javascript method call. This is similar to how we implement the `on` method. We can probably generalize method invocations further. Instead of create a new Branca element for each new method name, we can create a generalized `MethodInvocation` Branca element.
5. `MethodInvocations` are added to the Javascript code block.
6. HTML generation (`render`) is invoked on the Javascript code block.

**Comment by hansthen** [comment]
**Second part, make it backwards compatible**
1. Define a (thread local) variable referring to the Javascipt code block. This variable should be automatically initialized if it is not there.
2. Inside the Folium `Class` constructor, the object is added to the Javascript code block.
3. Inside the `MethodInvocation` constructor, the object is added to the Javascript code block.
4. Branca methods that currently exist on any folium objects are forwarded to the Javascript code block.

**Comment by Conengmo** [comment]
First I'll try to see if I understood correctly.

Seems like you define two goals:
- allow reusing elements (add an icon once, use it in multiple markers)
- allow circular referencing (references are not limited to the order in which objects are added to the map)

Backwards compatibility is only maintained for users of the documented Folium features, not for external plugins or users extending Folium in custom ways.

To understand the steps you are describing, it would really help me to see some code. Maybe you can make a very barebones example of this setup? Like, separately from the existing Folium code, just a dummy example to show the proposed architecture? It's interesting and I very much would like to understand it fully.

**Reply by hansthen** [reply]
Yes. A very concrete use case where I ran into a brick wall is the following. I have a Folium app that combines a LayerControl with a few Realtime layers. When a Realtime layer fails to load, I want to change the color of the name of that Layer inside the LayerControl. To do so, I want to add a reference to the LayerControl (via javascript options) when initializing the Realtime layer.

**Reply by Conengmo** [reply]
Question about this use case: if elements are rendered in order, how would a circular reference work? 

Either layer control doesn't know about realtime when it is initialized:

```
layer_control = LayerControl().add_to(m)  # this initializes with all existing layers
RealTime(color_option=layer_control)
```

or realtime doesn't konw about layer control yet:

```
RealTime(color_option=layer_control)  # error
layer_control = LayerControl().add_to(m)  # this initializes with all existing layers
```

**Comment by Conengmo** [comment]
I'm also inspired to try out adding multiple parents to the Branca element and still allow rendering. I.e. changing `Element._parent` to `Element._parents` and make sure each element is only rendered once. That would probably only address the first goal of reusing elements, but not allow circular referencing. I'll see if I can make a demo.

**Reply by Conengmo** [reply]
Adding code to deal with multiple parents in the graph is not that hard. The issue lies in how most of our templates in Folium currently are. The creation of the Javascript object is put together with adding it to a parent. To allow reuse, we would have to split that. First create all Javascript objects. Then add them together. 

Otherwise with the current templates you get either duplication of the same objects:

```
marker1 = L.marker()
icon = L.icon(my-icon)
marker1.setIcon(icon)

marker2 = L.marker()
icon = L.icon(my-icon)  # variable name clash
marker2.setIcon(icon)
```

or you get premature references

```
marker1 = L.marker()
marker1.setIcon(icon)  # does not exist here yet

marker2 = L.marker()
icon = L.icon(my-icon)
marker2.setIcon(icon)
```

**Reply by hansthen** [reply]
I agree we have to split the creation of Javascript objects and further handling of them. Both our proposals seem to do this somehow. I cannot immediately grok the consequences to Branca to allow elements to have multiple parents, and also how we would avoid elements being rendered multiple times.

**Reply by Conengmo** [reply]
> how we would avoid elements being rendered multiple times

I was thinking on render time we could collect some state on what elements are already rendered and what not.

But the real issue here is that templates both create objects and define relationships between objects, so we have to address that. Just making `_parents` a list is not going to solve it. That's where I stranded with this idea.

> Both our proposals seem to do this somehow

I don't really have a proposal for that so I'm interested in yours 😄

**Reply by hansthen** [reply]
> But the real issue here is that templates both create objects and define relationships between objects, so we have to address that. 

Indeed. We'd have to split the templates into the parts that create objects and the parts that define relationships between objects (or rather the parts that modify existing objects). We already do that in some places. E.g. `EventHandler`, `FitBounds`, `ElementToAddElement`, `JSCSSMixin`. We just make this consistent throughout Folium.

The root problem is 
1. We try to mimic the resulting javascript Leaflet object structure in our Branca parent child relationships.
2. We use the Branca parent child relationships to determine the order of rendering.

My suggestion is to get rid of 1. Instead of a recursive structure, all Folium elements have one global root element (the javascript block). In other words `self.get_root() == self.parent`.

So instead of this structure:

```
Root 
  Map
    LayerControl
      FeatureGroup
        Layer
``` 
```
Root
  Map # var map = new Map()
  LayerControl # var layerControl = new LayerControl()
  LayerControl.addTo(Map) # layerControl.addTo(map)
  FeatureGroup # var fg = new FeatureGroup() 
  Layer # var layer = new Layer() 
  Layer.addTo(FeatureGroup) # layer.addTo(fg)
  LayerControl.addTo(FeatureGroup) # fg.addTo(layerControl)
```

I realize this is a big change, with lots of tricky edge cases to be resolved during the change. However, I also think the new underlying structure will become much simpler so Folium will become easier to maintain.

**Reply by Conengmo** [reply]
Awesome, I get this idea. Makes a lot of sense. We should try it out to see where the issues lie!

**Reply by hansthen** [reply]
One of the issues I foresee is that Layer now has a `show` variable, specifically as a workaround for this issue. (So that an item can be added to the map in Folium without adding it to the map in Leaflet). 

I don't know what the best approach will be for those.

I have some time available during the holidays. I can write a small PoC and see where I get stuck.

**Reply by Conengmo** [reply]
One more disjointed thought: right now it's a bit confusing that from a user perspective `Map` is the root object. But underwater it adds a `Figure` as parent of `Map`. Maybe we can do without that, and `Map` can be the actual root that collects elements. Doesn't mean that all code has to live in the `Map` class, it could delegate that through composition.

**Reply by hansthen** [reply]
I was thinking about that, but what will we do with DualMap?

**Reply by Conengmo** [reply]
good point, would have to look into that!

**Comment by Conengmo** [comment]
Another thought: seems the Github Issues are just about adding icon and divicon to markers. Maybe the simplest thing is to just do something custom for those particular elements. That's not unheard of, we have other classes that do some special stuff around rendering.

**Reply by hansthen** [reply]
Yes, that is surely the most pragmatic approach. However, it does not solve the issue of circular references. I also ran in some other limitations due to the current structure: e.g. creating popups for geojson layers.

**Reply by Conengmo** [reply]
Created a prototype for this in https://github.com/python-visualization/folium/pull/2053. B.t.w. I don't mean this as saying we shouldn't look at other approaches.

Note that it's not meant to solve circular references or popups or tooltips. This is just for icons, since the thought was that all three issues we got we're explicitly about icons, so maybe that's the only thing we have to solve for.

**Comment by hansthen** [comment]
@Conengmo 

I made some progress for a PoC. There is one problem that I immediately ran into. I cannot easily use a thread local variable as the new root element. This will work in most typical use cases, but not for running the tests. However, we can probably use the Figure element that is used as the parent of `Map` to work as global root element.

There are however a few things I do not really understand in Branca that hinder my progress. How does the `render` method work in Branca? On `Element` the `render` method returns its generated template, which recursively calls `render` for each of its children. So far so good. When I get to `Figure` I see this:

```
      def render(self, **kwargs):
          """Renders the HTML representation of the element."""
          for name, child in self._children.items():
              child.render(**kwargs)
          return self._template.render(this=self, kwargs=kwargs)
```

Where does the output of the rendered children go? I also see that some render methods do not render anything at all, but just do structural manipulation (adding children to their parents and stuff). Am I correct that there are two versions of `render` 
1. that invokes a jinja template to generate html code.
2. one that prepares the parent to be rendered?

**Reply by Conengmo** [reply]
That's a great question. The way rendering works is very convoluted and still confuses me after all these years. At times I really dive into it and then feel like I understand, only for that knowledge to then leave my brain again.

I'll try if I can write a short step guide...

A thought I've also had is to simplify this part and move it back to Folium. That way this aspect would be easier to understand, and we can change it more easily without having to deal with separate releases.

**Reply by hansthen** [reply]
@Conengmo I just started a new discussion about this oddity, since it is also the thing that makes mypy complain. I think we should do something about this, since I have the exact same feeling that the truth is just a bit beyond my grasp. Can you have a look at it to see if my writeup makes sense? I also have a modest proposal to make it simpler.
